### ISSUE TITLE
Fix `declaration-property-value-no-unknown` false negatives for values containing known custom properties

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#9464]
### What minimal example or steps are needed to reproduce the bug?

```css
:root {
  --foo: red;
}

a {
  padding: var(--foo);
}
```

### What minimal configuration is needed to reproduce the bug?

```mjs
export default {
  rules: {
    "declaration-property-value-no-unknown": true,
  }
};
```

### How did you run Stylelint?

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcCBOEIBcAEwA6A7DGAtIQGarwZIxQDceAvnngIZZ4EAOzUUAlrgHMKAN2ZIAFMTIQAlHVz0QAGhAleAGxgA5ZgFs4iGAA89HTQDowAZyvLwEXGoEIQxjhCSZYJZgFd1mDj4lP4wVhRBBATYILBg6mLMaLwOhBwoHDCeAJ6Eouq+MIS4EIS+uADWJQDuuDEUaEiFSuwYjArydpCOvAIAYh66SS5WaNma6vxolg5O5roAVrYqsBy2iJEYMaPjMJO4aPVbIAloYYfKrdtjE1OE3U6Eo8y4UGJQRzGn5zEMeCD0IA)

### Which Stylelint-related dependencies are you using?

```json
{
  "stylelint": "17.14.1",
  "stylelint-config-standard": "40.0.0"
}
```

### What did you expect to happen?

The linter to report a problem for:

```css
padding: var(--foo);
```

As the value of `var(--foo)` is a `<color>` not a `<length>`.

### What actually happened?

No problems reported.

### Do you have a proposal to fix the bug?

We can substitute in the values of _known_ custom properties before the `lexer.matchProperty` check.

Before we added `referenceFiles`, most custom properties were unknown (unless they were defined in the same file). Now that most, if not all, are known, adding this check has become worthwhile.

We'd continue to ignore values that contain unknown custom properties.

I suggest we also add the substitution value as a new message arg, as we do with our resolved selectors, e.g.:

~~`Unknown value "var(--foo)" ("red") for property "padding"` (where `red` is the new arg).~~

`Unknown value "red" for property "padding", via "var(--foo)"`

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for the proposal. Substituting known custom properties seems feasible. 👍🏼 

After reviewing the request, I have the following points.

## The message wording

```
Unknown value "var(--foo)" ("red") for property "padding"
```

The message seems to read ambiguously: is `var(--foo)` unknown, or is `red`? And `var(--foo)` isn't unknown in any sense.

I'd suggest a separate message, as we already do for math with `rejectedMath`:

```
Invalid computed value "red" from "var(--foo)" for property "padding"
```

## False positives from the cascade

A custom property doesn't have one value. See examples:

```css
:root { --size: 10px; }
.compact { --size: 4px; }
```

We need to determine which one applies to the rule with some logic, or just give it up.

**Comment by jeddy3** [comment]
Two excellent points, thanks.

> I'd suggest a separate message, as we already do for math with `rejectedMath`:

I think this one indicates a problem with the rule's scope. The rule checks whether a value is "known" _to a specific property_.

However, it also currently checks:

- if a value is unparsable (flagged as invalid)
- if a math expression is invalid

Hence the two additional messages.

I wonder whether those checks belong in a separate `value-no-invalid` rule, since _neither depends on the property_. 

This would be a similar split to `selector-no-unmatchable` and `selector-no-invalid`.

>  seems to read ambiguously

We can then remove the ambiguity by appending `, via "$customProperty"` (as the [supplementary context](https://stylelint.io/contributor-guide/rules#add-problem-messages)) to the default problem message if there's a substitution. 

For example:

```
Unknown value "red" for property "padding", via "var(--foo)"
```

It's clear that `red` is unknown _to the `padding` property_. It also doesn't use "computed," as that refers to [something specific](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Cascade/Property_value_processing#computed_value).

I can open another issue to add the `value-no-invalid` rule if we think this direction has merit.

> False positives from the cascade

To avoid false positives, let's flag only when all definitions are unknown.

In your example, both are `<length>`s and known to the `padding` property, so it won't flag:

```css
:root { --size: 10px; }
.compact { --size: 4px; }
.foo { padding: var(--size); }
```

Nor will it flag if the types diverge, as we can't be sure of the cascade:

```css
:root { --size: red; }
.compact { --size: 4px; }
.foo { padding: var(--size); }
```

It will flag when all are unknown, though:

```css
:root { --size: 10px; }
.compact { --size: 4px; }
.foo { color: var(--size); }
```

In the problem message, we can display the last-in definition.

I believe that approach will avoid false positives while supporting the common use case of custom properties for theming.

**Comment by ybiquitous** [comment]
Great insights. I totally agree with splitting the current rule into a new *invalid* rule to eliminate the ambiguity. 👍🏼 

Also, the suggested approach for the cascade problem sounds reasonable. 👍🏼 

---

This is out of scope, but could we add a few words to the [Writing rules](https://stylelint.io/contributor-guide/rules) guide to remove the ambiguity between "invalid" and "unknown" in rule names/scopes?　Here's a rough idea:

> NOTE: A rule's concerns are often either context-dependent (e.g., "is this value unknown to this property?") or context-independent (e.g., "is this value syntactically invalid?"). These often justify separate rules.

**Comment by ybiquitous** [comment]
One more question: should this rule also consider a custom property defined by `@property`?

Example:

```css
@property --size { syntax: "<length>"; }

a { color: var(--size); }
```

**Comment by jeddy3** [comment]
> I totally agree with splitting the current rule into a new invalid rule

Nice. I've opened an issue for this:

- https://github.com/stylelint/stylelint/issues/9467

> should this rule also consider a custom property defined by `@property`

Yes, but only if they have an `initial-value` because we need a value to substitute in:

```css
@property --size { 
  syntax: "<length>";
  initial-value: 10px;
}

a { color: var(--size); }
```

Alternatively, we can provide our own substitutions for `@property` definitions that omit `initial-value`. There are only a [dozen or so](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/At-rules/@property/syntax) valid values for `syntax`. However, we'd need to leave out the value in the problem message; otherwise, it'll be confusing for people:

```
Unknown value for property "color", via "var(--size)"
```

I suggest we start with false negatives for `@property`s without `initial-value`.

**Comment by ybiquitous** [comment]
> However, we'd need to leave out the value in the problem message

Sounds good. The value might make the message unexpectedly long. 👍🏼
